### PR TITLE
Custom transform support

### PR DESCRIPTION
**Pull request by Sentient07**
This PR aims to change the `FasterRCNN` class to accept custom Transformations  from the user, foregoing the generic `GeneralizedRCNNTransform` as the compulsory transform.

### CODE REVIEW

File: torchvision/models/detection/generalized_rcnn.py [hunk 1]
```diff
@@ -97,7 +99,7 @@ def forward(self, images, targets=None):
             features = OrderedDict([('0', features)])
         proposals, proposal_losses = self.rpn(images, features, targets)
         detections, detector_losses = self.roi_heads(features, proposals, images.image_sizes, targets)
-        detections = self.transform.postprocess(detections, images.image_sizes, original_image_sizes)
+        detections = postprocess(detections, images.image_sizes, original_image_sizes, self.training)
```
Comment (fmassa) [inline torchvision/models/detection/generalized_rcnn.py:102]: I see this as a limitation, as now users can't customize the postprocessing themselves, while before it was possible to replace the transform with `model.transform = MyCustomTransform()` and it would work

File: torchvision/models/detection/faster_rcnn.py [hunk 2]
```diff
@@ -142,8 +142,10 @@ class FasterRCNN(GeneralizedRCNN):
 
     def __init__(self, backbone, num_classes=None,
                  # transform parameters
+                 transform=None,
                  min_size=800, max_size=1333,
-                 image_mean=None, image_std=None,
+                 image_mean=[0.485, 0.456, 0.406],
+                 image_std=[0.229, 0.224, 0.225],
```
Comment (fmassa) [inline torchvision/models/detection/faster_rcnn.py:148]: Can you revert those changes? It's in general preferable not to pass mutable objects as arguments to functions.

File: torchvision/models/detection/faster_rcnn.py [hunk 3]
```diff
@@ -224,11 +226,8 @@ def __init__(self, backbone, num_classes=None,
             bbox_reg_weights,
             box_score_thresh, box_nms_thresh, box_detections_per_img)
 
-        if image_mean is None:
-            image_mean = [0.485, 0.456, 0.406]
-        if image_std is None:
-            image_std = [0.229, 0.224, 0.225]
-        transform = GeneralizedRCNNTransform(min_size, max_size, image_mean, image_std)
+        if transform is None:
+            transform = GeneralizedRCNNTransform(min_size, max_size, image_mean, image_std)
```
Comment (fmassa) [inline torchvision/models/detection/faster_rcnn.py:230]: My comment before corresponds to the case where `transform is not None` but `min_size is not None` (which we should change the default to be `None`)
We should raise an error in this case.

Something in the lines of
```python
if transform is None:
    if min_size is None:
        min_size = 800
    ...
else:
    if min_size is not None:
        raise ValueError("Can't specify both transform and min_size")
```
The same for `max_size`, `image_mean` and `image_std`.

The reason why it would be good to make this check is that the user could have a very different type of custom transform, for which `max_size` is non-existing. In this case, specifying a `max_size` would just be confusing, so we should either use `transform` or the default parameters.

File: torchvision/models/detection/faster_rcnn.py [hunk 4]
```diff
@@ -142,8 +142,10 @@ class FasterRCNN(GeneralizedRCNN):
 
     def __init__(self, backbone, num_classes=None,
                  # transform parameters
+                 transform=None,
```
Comment (fmassa) [inline torchvision/models/detection/faster_rcnn.py:145]: This is in theory BC-breaking, as users could have created the model with ordered arguments, which will fail now.

Could you put it instead in the end?

File: torchvision/models/detection/generalized_rcnn.py (hunk 1 above)
Comment (Sentient07) [reply torchvision/models/detection/generalized_rcnn.py:102]: Yes! You're right. I didn't think out of the bounding box's perspective and indeed it's different for mask and keypoints, for which user choice seem to be a nice option

**Comment by Sentient07** [comment]
Hi @fmassa I've made the changes you've suggested. Let me know what you think. I couldn't test it locally again due to the previously mentioned error. I'm sorry for burdening the CI. Once the builds finish, I will check again if there are errors and fix them
